server/master/src/main/java/org/apache/accumulo/master/TimeoutTaskExecutor.java

```diff
@@ -0,0 +1,235 @@
+   * @throws IllegalArgumentException If numThreads is less than 1 or expectedNumCallables is negative.
+   */
+  public TimeoutTaskExecutor(int numThreads, long timeout, int expectedNumCallables) {
+    Preconditions.checkArgument(numThreads >= 1, "Number of threads must be at least 1.");
+    Preconditions.checkArgument(expectedNumCallables >= 0, "The expected number of callables must be non-negative.");
+
+    this.executorService = Executors.newFixedThreadPool(numThreads);
+    this.startedTasks = new ArrayBlockingQueue<>(expectedNumCallables);
+    this.timeout = timeout;
+    this.wrappedTasks = new ArrayList<>(expectedNumCallables);
+  }
+
+  /**
+   * Submits a new task to the executor.
+   *
+   * @param callable Task to run
+   */
+  public void submit(C callable) {
+    WrappedTask wt = new WrappedTask(callable);
+    wt.future = executorService.submit(wt);
+    wrappedTasks.add(wt);
+  }
+
+  /**
+   * Registers the callback to use on successful tasks.
+   *
+   * @param successCallback The callback function to invoke on success.
+   * @throws NullPointerException when a null successCallback is passed in
+   */
+  public void onSuccess(SuccessCallback<T,C> successCallback) {
+    this.successCallback = Objects.requireNonNull(successCallback, "Must provide a non-null successCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that throw exceptions.
+   *
+   * @param exceptionCallback The callback function to invoke on exceptions.
+   * @throws NullPointerException when a null exceptionCallback is passed in
+   */
+  public void onException(ExceptionCallback<C> exceptionCallback) {
+    this.exceptionCallback = Objects.requireNonNull(exceptionCallback, "Must provide a non-null exceptionCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that time out.
+   *
+   * @param timeoutCallback The callback function to invoke on timeouts.
+   * @throws NullPointerException when a null timeoutCallback is passed in
+   */
+  public void onTimeout(TimeoutCallback<C> timeoutCallback) {
+    this.timeoutCallback = Objects.requireNonNull(timeoutCallback, "Must provide a non-null timeoutCallback.");
+  }
+
+  /**
+   * Completes all the current tasks by dispatching to the appropriate callback.
+   *
+   * @throws IllegalStateException If all of the callbacks were not registered before calling this method.
+   * @throws InterruptedException  If interrupted while awaiting callable results.
+   */
+  public void complete() throws InterruptedException {
+    Preconditions.checkState(successCallback != null, "Must set a success callback before completing " + this);
+    Preconditions.checkState(exceptionCallback != null, "Must set an exception callback before completing " + this);
+    Preconditions.checkState(timeoutCallback != null, "Must set a timeout callback before completing " + this);
+
+    while (hasUnfinishedTasks()) {
+      completeTask(startedTasks.take());
```

Comment:
When I looked at this code, I asked myself "Can the thread ever get stuck in take() forever?"  I can not see how this could happen, however I have not been able to prove to myself that it can't happen.  Could switch this loop to poll for a limited time (like a second), but I am not sure if this needs to be done. I have always wanted a way to close a blocking queue.

Reply:
I know what you mean. I can make it a poll(timeout) as a safety/sanity catch.